File: java/src/main/java/io/cucumber/query/LineageReducerAscending.java
package io.cucumber.query;

import io.cucumber.messages.types.Pickle;

import java.util.function.Supplier;

import static java.util.Objects.requireNonNull;

/**
 * Reduces the lineage of a Gherkin document element in ascending order.
 *
 * @param <T> type to which the lineage is reduced.
 */
class LineageReducerAscending<T> implements LineageReducer<T> {

    private final Supplier<? extends Collector<T>> collectorSupplier;

    LineageReducerAscending(Supplier<? extends Collector<T>> collectorSupplier) {
        this.collectorSupplier = requireNonNull(collectorSupplier);
    }

    @Override
    public T reduce(Lineage lineage) {
        Collector<T> collector = collectorSupplier.get();
        reduceAddLineage(collector, lineage);
        return collector.finish();
    }

    @Override
    public T reduce(Lineage lineage, Pickle pickle) {
        Collector<T> collector = collectorSupplier.get();
        collector.add(pickle);
        reduceAddLineage(collector, lineage);
        return collector.finish();
    }

    private static <T> void reduceAddLineage(Collector<T> reducer, Lineage lineage) {
        lineage.example().ifPresent(example -> reducer.add(example, lineage.exampleIndex().orElse(0)));
        lineage.examples().ifPresent(examples -> reducer.add(examples, lineage.examplesIndex().orElse(0)));
        lineage.scenario().ifPresent(reducer::add);
        lineage.rule().ifPresent(reducer::add);
        lineage.feature().ifPresent(reducer::add);
        reducer.add(lineage.document());
    }
}

File: dotnet/Query/LineageReducerAscending.cs
using System;
using Io.Cucumber.Messages.Types;

namespace Io.Cucumber.Query
{
    public class LineageReducerAscending<T> : ILineageReducer<T>
    {
        private readonly Func<ICollector<T>> _collectorSupplier;

        public LineageReducerAscending(Func<ICollector<T>> collectorSupplier)
        {
            _collectorSupplier = collectorSupplier ?? throw new ArgumentNullException(nameof(collectorSupplier));
        }

        public T Reduce(Lineage lineage)
        {
            var collector = _collectorSupplier();
            ReduceAddLineage(collector, lineage);
            return collector.Finish();
        }

        public T Reduce(Lineage lineage, Pickle pickle)
        {
            var collector = _collectorSupplier();
            collector.Add(pickle);
            ReduceAddLineage(collector, lineage);
            return collector.Finish();
        }

        private static void ReduceAddLineage(ICollector<T> collector, Lineage lineage)
        {
            if (lineage.Example != null)
                collector.Add(lineage.Example, lineage.ExampleIndex ?? 0);
            if (lineage.Examples != null)
                collector.Add(lineage.Examples, lineage.ExamplesIndex ?? 0);
            if (lineage.Scenario != null)
                collector.Add(lineage.Scenario);
            if (lineage.Rule != null)
                collector.Add(lineage.Rule);
            if (lineage.Feature != null)
                collector.Add(lineage.Feature);
            if (lineage.Document != null)
                collector.Add(lineage.Document);
        }
    }
}
